### PR TITLE
Add empty to State and StateT objects

### PR DESCRIPTION
**Pull request by larsrh**
Work by @sderosiaux in #2294. Rebased to current master. For some reason, the bincompat error also went away (at least locally).

### CODE REVIEW

**Comment by LukaJCB** [comment]
Seems like there's still a bincompat issue for 2.11

**Comment by larsrh** [comment]
Ah, right. 2.11 doesn't like methods added to traits. But this should be fine with an exception, no?

**Comment by LukaJCB** [comment]
We can add new methods by creating a new trait and mixing it in, similar to what we've been doing with the syntax additions :)

**Comment by larsrh** [comment]
Smart. I've updated the commit accordingly.

**Comment by larsrh** [comment]
And we're green!

File: core/src/main/scala/cats/data/IndexedStateT.scala [hunk 1]
```diff
@@ -223,6 +223,11 @@ object IndexedStateT extends IndexedStateTInstances with CommonStateTConstructor
     IndexedStateT(_ => F.map(fsb)(s => (s, ())))
 }
 
+private[data] trait CommonStateTConstructors0 extends CommonStateTConstructors {
+  def empty[F[_], S, A: Monoid](implicit F: Applicative[F]): IndexedStateT[F, S, S, A] =
+    pure(Monoid[A].empty)
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/IndexedStateT.scala]: Is there a reason to use `: Monoid` for `A` but not use `: Applicative` for `F`? I'd be inclined to not use the syntactic sugar for either so you can avoid do something like `F.pure(A.empty)` and avoid an extra `Monoid.apply[A]` call.

**Comment by larsrh** [comment]
@ceedubs Is this fine with you now?

File: core/src/main/scala/cats/data/IndexedStateT.scala [hunk 2]
```diff
@@ -223,6 +223,11 @@ object IndexedStateT extends IndexedStateTInstances with CommonStateTConstructor
     IndexedStateT(_ => F.map(fsb)(s => (s, ())))
 }
 
+private[data] trait CommonStateTConstructors0 extends CommonStateTConstructors {
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/IndexedStateT.scala]: Is there any particular reason to introduce this new trait as opposed to just putting this method directly into the `object`?

File: core/src/main/scala/cats/data/IndexedStateT.scala [hunk 3]
```diff
@@ -299,6 +304,11 @@ private[data] abstract class StateFunctions {
    */
   def pure[S, A](a: A): State[S, A] = State(s => (s, a))
 
+  /**
+    * Return `A`'s empty monoid value and maintain the input state.
+    */
+  def empty[S, A: Monoid]: State[S, A] = pure(Monoid[A].empty)
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/IndexedStateT.scala]: Should this use the same approach used above for getting the `empty` of the `Monoid`?

File: core/src/main/scala/cats/data/IndexedStateT.scala (hunk 2 above)
Comment (larsrh) [reply core/src/main/scala/cats/data/IndexedStateT.scala]: Binary compatibility: https://github.com/typelevel/cats/pull/2533#issuecomment-424601003

Comment (larsrh) [reply core/src/main/scala/cats/data/IndexedStateT.scala]: To be precise, the function is not only provided on the `IndexedStateT` object, but also on `StateT`. This is the reason why it has to go into a trait in the first place.

**Comment by larsrh** [comment]
@ceedubs Updated accordingly.

File: core/src/main/scala/cats/data/IndexedStateT.scala (hunk 2 above)
Comment (ceedubs) [reply core/src/main/scala/cats/data/IndexedStateT.scala]: Oh okay that makes sense. Thanks for the explanation.

**Pull request merged**